### PR TITLE
Fix support for queries containing sort operators on JDBC data stores

### PR DESCRIPTION
**Pull request by vogti**
This PR fixes support for sort-operations on relational data stores. These most probably got broken in context of the restructuring of the whole query processing. Fixing this required additional changes to the query processing itself.

Together with this PR, the following minor aspects get addressed as well:

- Jdbc Sort operators should now be pushed down more reliably.
- Generated code should now always be displayed in the monitoring view of the UI.
- Minor improvements and tweaks to the query processing.
- Some refactoring and code clean-up.

### CODE REVIEW

File: webui/build.gradle [hunk 1]
```diff
@@ -62,8 +62,6 @@ dependencies {
 
     implementation group: "com.sparkjava", name: "spark-core", version: spark_core_version
 
-    implementation group: "ch.unibas.dmi.dbis.polyphenydb", name: "polyphenydb-jdbc-driver", version: polyphenydb_jdbc_driver_version
```
Comment (schoenja) [inline webui/build.gradle]: Has this been replaced by another dependency? Or is it simply no longer needed?

File: dbms/src/main/resources/log4j2.xml [hunk 2]
```diff
@@ -44,6 +44,10 @@
         </logger>
         <!-- <logger name="ch.unibas.dmi.dbis.polyphenydb.jdbc" level="TRACE">
         </logger> -->
+        <!--<logger name="ch.unibas.dmi.dbis.polyphenydb.plan.RelOptPlanner" level="DEBUG">
+        </logger> -->
+        <!-- <logger name="ch.unibas.dmi.dbis.polyphenydb.plan.hep.HepPlanner" level="TRACE">
+        </logger> -->
```
Comment (schoenja) [inline dbms/src/main/resources/log4j2.xml]: Why are there commented out lines added here?

File: core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/adapter/enumerable/EnumerableInterpretable.java [hunk 3]
```diff
@@ -107,32 +108,35 @@ public EnumerableInterpretable copy( RelTraitSet traitSet, List<RelNode> inputs
...
         String s = Expressions.toString( expr.memberDeclarations, "\n", false );
 
-        if ( PolyphenyDbPrepareImpl.DEBUG ) {
+        if ( RuntimeConfig.DEBUG.getBoolean() ) {
             Util.debugCode( System.out, s );
         }
 
-        if ( dataContext != null && dataContext.getTransaction().isAnalyze() ) {
-            InformationManager queryAnalyzer = dataContext.getTransaction().getQueryAnalyzer();
+        if ( transaction.isAnalyze() ) {
```
Comment (schoenja) [inline core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/adapter/enumerable/EnumerableInterpretable.java]: Do we know for sure that `transaction` is never `null`? If not, should we keep the null check?

File: dbms/src/main/java/ch/unibas/dmi/dbis/polyphenydb/processing/VolcanoQueryProcessor.java [hunk 4]
```diff
@@ -55,6 +77,63 @@
...
+                    ReduceExpressionsRule.PROJECT_INSTANCE,
+                    ReduceExpressionsRule.FILTER_INSTANCE,
+                    ReduceExpressionsRule.CALC_INSTANCE,
+                    ReduceExpressionsRule.JOIN_INSTANCE,
+                    ValuesReduceRule.FILTER_INSTANCE,
+                    ValuesReduceRule.PROJECT_FILTER_INSTANCE,
+                    ValuesReduceRule.PROJECT_INSTANCE,
+                    AggregateValuesRule.INSTANCE );
+
+
```
Comment (schoenja) [inline dbms/src/main/java/ch/unibas/dmi/dbis/polyphenydb/processing/VolcanoQueryProcessor.java]: Could you explain why this changes was made?

File: core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/rex/RexExecutorImpl.java [hunk 5]
```diff
@@ -124,9 +128,18 @@ private String compile( RexBuilder rexBuilder, List<RexNode> constExps, RexToLix
...
         }
+        if ( dataContext != null && dataContext.getTransaction() != null && dataContext.getTransaction().isAnalyze() ) {
+            InformationManager queryAnalyzer = dataContext.getTransaction().getQueryAnalyzer();
+            InformationPage page = new InformationPage( "informationPageGeneratedCode", "Generated Code" );
+            InformationGroup group = new InformationGroup( page, "Generated Code" );
+            queryAnalyzer.addPage( new InformationPage( "informationPageGeneratedCode", "Generated Code" ) );
+            queryAnalyzer.addGroup( group );
+            InformationCode informationCode = new InformationCode( group, code );
+            queryAnalyzer.registerInformation( informationCode );
+        }
```
Comment (schoenja) [inline core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/rex/RexExecutorImpl.java]: Please explain this change, what it does, in the Pull Request.

File: jdbc-interface/src/main/java/ch/unibas/dmi/dbis/polyphenydb/jdbc/DbmsMeta.java [hunk 6]
```diff
@@ -981,7 +988,9 @@ public ExecuteResult prepareAndExecute( final StatementHandle h, final String sq
                         h.id,
                         false,
                         signature,
-                        maxRowsInFirstFrame > 0 ? fetch( h, 0, (int) Math.min( Math.max( maxRowCount, maxRowsInFirstFrame ), Integer.MAX_VALUE ) ) : Frame.MORE // Send first frame to together with the response to save a fetch call
+                        maxRowsInFirstFrame > 0 && SEND_FIRST_FRAME_WITH_RESPONSE
+                                ? fetch( h, 0, (int) Math.min( Math.max( maxRowCount, maxRowsInFirstFrame ), Integer.MAX_VALUE ) )
+                                : null //Frame.MORE // Send first frame to together with the response to save a fetch call
```
Comment (schoenja) [inline jdbc-interface/src/main/java/ch/unibas/dmi/dbis/polyphenydb/jdbc/DbmsMeta.java]: In the inline else case, this now returns `null` instead of `Frame.MORE`. Is this intended behaviour?
Additionally it seems that this if will always evaluate to false, and execute the else branch.

File: webui/build.gradle (hunk 1 above)
Comment (vogti) [reply webui/build.gradle]: This is already since quite some time not required anymore.

File: dbms/build.gradle [hunk 7]
```diff
@@ -81,7 +81,7 @@ dependencies {
     //testCompile group: "net.hydromatic", name: "quidem", version: quidem_version
     //testCompile group: "net.hydromatic", name: "scott-data-hsqldb", version: scott_data_hsqldb_version
 
-    testCompile group: "ch.unibas.dmi.dbis.polyphenydb", name: "polyphenydb-jdbc-driver", version: polyphenydb_jdbc_driver_version
+    testCompile group: "org.polypheny", name: "polypheny-jdbc-driver", version: polypheny_jdbc_driver_version
```
[the remark this answers is not in the document]
Comment (vogti) [reply dbms/build.gradle]: This will happen with all other parts as well.

File: dbms/src/main/resources/log4j2.xml (hunk 2 above)
Comment (vogti) [reply dbms/src/main/resources/log4j2.xml]: Because this is helpful for debugging the planner and optimizer. Leaving them in the file allows to easily enable them if required.

File: core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/adapter/enumerable/EnumerableInterpretable.java (hunk 3 above)
Comment (vogti) [reply core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/adapter/enumerable/EnumerableInterpretable.java]: There was no null check for transaction, only for dataContext. But you are right: a null check won't hurt. I will add one.

File: dbms/src/main/java/ch/unibas/dmi/dbis/polyphenydb/processing/VolcanoQueryProcessor.java (hunk 4 above)
Comment (vogti) [reply dbms/src/main/java/ch/unibas/dmi/dbis/polyphenydb/processing/VolcanoQueryProcessor.java]: Simply refactoring. These rules have before been defined in another class.

File: jdbc-interface/src/main/java/ch/unibas/dmi/dbis/polyphenydb/jdbc/DbmsMeta.java (hunk 6 above)
Comment (vogti) [reply jdbc-interface/src/main/java/ch/unibas/dmi/dbis/polyphenydb/jdbc/DbmsMeta.java]: Yes, this is intended and necessary due to a bug in Avatica.

Comment (vogti) [reply jdbc-interface/src/main/java/ch/unibas/dmi/dbis/polyphenydb/jdbc/DbmsMeta.java]: Why do you think it always evaluated to false?

Comment (schoenja) [reply jdbc-interface/src/main/java/ch/unibas/dmi/dbis/polyphenydb/jdbc/DbmsMeta.java]: `SEND_FIRST_FRAME_WITH_RESPONSE` is always false, so we get:

`maxRowsInFirstFrame > 0 && SEND_FIRST_FRAME_WITH_RESPONSE` -> `maxRowsInFirstFrame > 0 && false` -> `false`

Unless `SEND_FIRST_FRAME_WITH_RESPONSE` is modified elsewhere, which would be a bad design.

Comment (vogti) [reply jdbc-interface/src/main/java/ch/unibas/dmi/dbis/polyphenydb/jdbc/DbmsMeta.java]: Sorry, I misunderstood your comment. Yes, this is correct and intended. Due to a bug in avatica it fails under some circumstances when sending data together with the first frame. So I "disabled" that, but left the code to have a closer look on that in future.

**Comment by schoenja** [comment]
> Thx @schoenja for the review!
> 
> > There were a lot of changes in regards to the order of imports. Were those intended?
> 
> These changes are either due to refactoring or have been made by my IDE.
> 
Okay great, just wanted to make sure that it wasn't the IDE going rogue and doing something that wasn't intended to happen. Maybe we should run the reformatting over the entire project so that the entire codebase is properly formatted?

> > Please give a more thorough overview over the things that were changed in the pull request, as it will be referenced by the merge commit, and should help others understand what the changes do, why they were done.
> 
> What do you think should be elaborated in more detail?
> 
I think a short summary of what the issue was, a brief overview of how it was fixed, and a short explanation of exactly what `minor improvements and tweaks to the query processing.` are would be helpful.
> > The pull request title doesn't fully convey the changes in this PR
> 
> The title states the reason for this PR. What do you miss?
>
Rereading it, I don't know why I wrote this.

File: jdbc-interface/src/main/java/ch/unibas/dmi/dbis/polyphenydb/jdbc/DbmsMeta.java (hunk 6 above)
Comment (schoenja) [reply jdbc-interface/src/main/java/ch/unibas/dmi/dbis/polyphenydb/jdbc/DbmsMeta.java]: Then I think this behaviour should be documented with the code, possibly including a `fixme` or `todo` comment.

Comment (vogti) [reply jdbc-interface/src/main/java/ch/unibas/dmi/dbis/polyphenydb/jdbc/DbmsMeta.java]: Yes, this is right. I will add a todo comment

File: core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/adapter/enumerable/EnumerableInterpretable.java (hunk 3 above)
Comment (vogti) [reply core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/adapter/enumerable/EnumerableInterpretable.java]: There was no null check for transaction, only for dataContext. But you are right: a null check won't hurt. I will add one.

File: core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/rex/RexExecutorImpl.java (hunk 5 above)
Comment (vogti) [reply core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/rex/RexExecutorImpl.java]: This is only for debugging purposes, but I will add a comment.
